src/java.desktop/macosx/classes/com/apple/laf/AquaMenuPainter.java

```diff
@@ -294,6 +295,10 @@ protected void paintMenuItem(final Client client, final Graphics g, final JCompo
 
         // Paint the Arrow
         if (arrowIcon != null) {
+            if (!isEnabled && (arrowIcon instanceof ImageIcon)) {
+                arrowIcon = new ImageIconUIResource(GrayFilter.
+                        createDisabledImage(((ImageIcon)arrowIcon).getImage()));
+            }
```

Comment:
In MenuArrowIcon, the arrow icon is drawn there itself using drawPolygon but in Aqua, it is drawn via imageicon image so it will be a change in design and will make it more complex and it might introduce bug in hidpi/retina. I will like to keep it simple following the current design in aqua.

Reply:
On windows it is painted as via skin, I guess in GTK as well. But it really does not matter how it is drawn, the important part is that the icon itself knows how to draw proper appearance.

Reply:
In other L&Fs, it creates their own icon 
for ex, MetalLookAndFeel
"MenuItem.arrowIcon",(LazyValue) t -> MetalIconFactory.getMenuItemArrowIcon(),
and WindowsLookAndFeel
"MenuItem.arrowIcon", menuItemArrowIcon,

but for AquaL&F it does not create it's own factory icon and rely on BasicIconFactory$MenuItemArrowIcon, so tweaking aqua arrow icon inline with how other do will cause design change, so I think this fix caters to the problem in view of this aqua limitation.

Reply:
> but for AquaL&F it does not create it's own factory icon and rely on BasicIconFactory$MenuItemArrowIcon, so tweaking aqua arrow icon inline with how other do will cause design change, so I think this fix caters to the problem in view of this aqua limitation.

Are you sure that the Aqua does not have its own factory? the BasicIconFactory$MenuItemArrowIconin is an empty class.
AquaLookAndFeel.java:
    "Menu.arrowIcon",(LazyValue) t -> AquaImageFactory.getMenuArrowIcon(),

Reply:
That is for MenuArrowIcon which is also there for Metal and Windows L&F in addition with MenuItemArrowIcon(). Aqua does not have it for MenuItem.
The control for aqua menuitem goes via ImageIcon#paintIcon for menuitem arrow which does not have enabled/disabled parameter so it was done the way it is now.

Reply:
> That is for MenuArrowIcon which is also there for Metal and Windows L&F in addition with MenuItemArrowIcon(). Aqua does not have it for MenuItem.

But this icon is used for the menuitem as well, isn't it? You can add this line to the test and check:
     UIManager.setLookAndFeel("com.apple.laf.AquaLookAndFeel");
+   UIManager.getDefaults().put("Menu.arrowIcon", "");

> The control for aqua menuitem goes via ImageIcon#paintIcon for menuitem arrow which does not have enabled/disabled parameter so it was done the way it is now.

I guess it is go via Icon.paintIcon() at AquaMenuPainter:408 , this is similar to metal and windows L&Fs. The first parameter of paintIcon is a component, so all information can be read there. And depending on the state that icon can draws default arrow, or creates/caches the disabled icon.

Reply:
> 
> 
> That is for MenuArrowIcon which is also there for Metal and Windows L&F in addition with MenuItemArrowIcon(). Aqua 

> And depending on the state that icon can draws default arrow, or creates/caches the disabled icon.

I guess that is what it is being done, no? Instead of inside paintArrow() method, I am creating the disabled icon outside.